### ISSUE TITLE
message not received by the server but no exception thrown

### ISSUE THREAD

**Issue by rbro** [php-amqplib/php-amqplib#595]
I am running into an issue where basic_publish runs with no errors, but the message is never received by the rabbitmq server.  The process is a long running process, so I'm calling check_heartbeat periodically to keep the connection going.

In the below example:
- message1 and message2 are received by the rabbitmq server
- message3 appears to be sent as no exceptions are thrown, but the rabbitmq server does not receive this message
- an exception is thrown for message4

Can you please check why message3 was not received by the server, yet didn't throw any exception either?

Thanks for your help.

```
<?php
require_once __DIR__.'/vendor/autoload.php';

class custom_connection extends \PhpAmqpLib\Connection\AMQPStreamConnection
{
	public function check_heartbeat()
	{
		$this->getIO()->check_heartbeat();
	}
}

$connection = new custom_connection('localhost', 5672, 'guest', 'guest', '/', $insist=false, $login_method='AMQPLAIN', $login_response=null, $locale='en_US', $connection_timeout=3, $read_write_timeout=60, $context=null, $keepalive=false, $heartbeat=2);

$channel = $connection->channel();

$channel->queue_declare('task_queue1', false, true, false, false);

for ($i = 1; $i <= 10; $i++)
{
	$connection->check_heartbeat();

	$message = 'message'.$i;

	echo $message."\n";

	$channel->basic_publish(new PhpAmqpLib\Message\AMQPMessage($message), '', 'task_queue1');

	sleep(4);
}
```

**Comment by lukebakken** [comment]
Hello,

Thanks for using this library and RabbitMQ.

There's not nearly enough information for someone to help out. Please provide the following:

* RabbitMQ version, Erlang version, operating system
* RabbitMQ configuration
* Relevant RabbitMQ logs, or the entire log file
* An *entire* working script to reproduce your issue.
* Details on how to run the script, and how you are running your script. If the above is your complete script, please say so and give instructions on how to run it.
* A TCP capture from your environment showing the issue (port 5672)

Thanks

**Comment by rbro** [comment]
Responses are below, thanks.  Please let me know if I can provide more info.

- RabbitMQ version: 3.7.7
- Erlang version: 21.0.5
- Operating System: CentOS 7.5
- PHP version: 7.2.10
- php-amqplib version: 2.7.2
- I'm using the default configuation provided by the RPM's on http://www.rabbitmq.com/install-rpm.html.
- I attached the RabbitMQ log file in [rabbitmq.txt](https://github.com/php-amqplib/php-amqplib/files/2386964/rabbitmq.txt).
- The entire client script is above.  I pasted a receive script below.
- Save the below script as receive.php and run it with "php receive.php" from the command-line.  Then save the above script as send.php and run it with "php send.php" from the command-line.
- I attached a tcpdump in [tcpdump.txt](https://github.com/php-amqplib/php-amqplib/files/2386965/tcpdump.txt) by running "tcpdump -vv -x -X -s 1500 -i lo 'port 5672'".  Please let me know if you need it with other options.

receive.php:
```
<?php
require_once __DIR__.'/vendor/autoload.php';

$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection('localhost', 5672, 'guest', 'guest');

$channel = $connection->channel();

$channel->queue_declare('task_queue1', false, true, false, false);

echo " [*] Waiting for messages. To exit press CTRL+C\n";

$callback = function ($msg)
{
    echo ' [x] Received ', $msg->body, "\n";
};

$channel->basic_consume('task_queue1', '', false, true, false, false, $callback);

while (count($channel->callbacks))
{
    $channel->wait();
}

$channel->close();

$connection->close();
```

**Comment by rbro** [comment]
Also, here is the output I get.  Again, message3 looks to be successfully sent, but receive.php never receives it.

receive.php:
```
 [*] Waiting for messages. To exit press CTRL+C
 [x] Received message1
 [x] Received message2
```

send.php:
```
message1
message2
message3
message4

Fatal error: Uncaught ErrorException: fwrite(): send of 66 bytes failed with errno=104 Connection reset by peer in /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/StreamIO.php:282
Stack trace:
#0 [internal function]: PhpAmqpLib\Wire\IO\StreamIO->error_handler(8, 'fwrite(): send ...', '/tmp/rabbitmq/v...', 282, Array)
#1 /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/StreamIO.php(282): fwrite(Resource id #29, '\x01\x00\x01\x00\x00\x00\x14\x00<\x00(\x00\x00\x00\v...', 8192)
#2 /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(341): PhpAmqpLib\Wire\IO\StreamIO->write('\x01\x00\x01\x00\x00\x00\x14\x00<\x00(\x00\x00\x00\v...')
#3 /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Connection/AbstractConnection.php(382): PhpAmqpLib\Connection\AbstractConnection->write('\x01\x00\x01\x00\x00\x00\x14\x00<\x00(\x00\x00\x00\v...')
#4 /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Channel/A in /tmp/rabbitmq/vendor/php-amqplib/php-amqplib/PhpAmqpLib/Wire/IO/StreamIO.php on line 282
```

**Comment by lukebakken** [comment]
I haven't had a chance to try this out, but what is the result when you remove the `sleep()` call from `send.php`?

**Comment by rbro** [comment]
If I remove the `sleep` call, then `send.php` runs successfully and all 10 messages are picked up by `receive.php`.

**Comment by lukebakken** [comment]
OK that's what I expected. There may be a way to "sleep" via the connection but continue processing data. I'll look into that, but you can check the code as well if you have time.

**Comment by rbro** [comment]
Yes, thanks.  I think it's something with hitting the heartbeat.  Also, just to be clear, in my actual application, I'm not using `sleep`.  I purposely set the heartbeat low to 2 seconds and was just using `sleep` to simulate doing work.  The structure in my actual application is more like:

```
<?php
// connect to rabbitmq

// queue message

// do work over many seconds

// check_heartbeat

// queue message

// repeat
```

**Comment by rbro** [comment]
It seems to be related to reconnect, and nothing related to heartbeat and sleep.  In the below send2.php, only message1 gets received by receive.php.  message2 through message10 do not get received.

send2.php:
```
<?php
require_once __DIR__.'/vendor/autoload.php';

class custom_connection extends \PhpAmqpLib\Connection\AMQPStreamConnection
{
	public function reconnect()
	{
		$this->getIO()->reconnect();
	}
}

$connection = new custom_connection('localhost', 5672, 'guest', 'guest', '/');

$channel = $connection->channel();

$channel->queue_declare('task_queue1', false, true, false, false);

for ($i = 1; $i <= 10; $i++)
{
	$message = 'message'.$i;

	echo $message."\n";

	$channel->basic_publish(new PhpAmqpLib\Message\AMQPMessage($message), '', 'task_queue1');

	$connection->reconnect();
}
```

**Comment by rbro** [comment]
From my testing, it appears to be an issue with the reconnect function in the StreamIO class.  If I call reconnect in the AbstractConnection class and then request a new channel, then all works as expected - see script below.

I'm not familiar with the internals, but does calling reconnect in the StreamIO class have different effects vs calling reconnect in AbstractConnection?  The check_heartbeat function in StreamIO calls reconnect in StreamIO, which is creating the issue.  Should check_heartbeat be calling reconnect in AbstractConnection instead?

Thanks for your help.

The below script works and all 10 messages are received since I called reconnect from AbstractConnection rather than StreamIO:

send3.php:
```
<?php
require_once __DIR__.'/vendor/autoload.php';

$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection('localhost', 5672, 'guest', 'guest', '/');

$channel = $connection->channel();

$channel->queue_declare('task_queue1', false, true, false, false);

for ($i = 1; $i <= 10; $i++)
{
	$message = 'message'.$i;

	echo $message."\n";

	$channel->basic_publish(new PhpAmqpLib\Message\AMQPMessage($message), '', 'task_queue1');

	$connection->reconnect();
	$channel = $connection->channel();
}
```

**Comment by rbro** [comment]
@hairyhum - thanks, I confirmed it now throws an exception after the heartbeat expires.  Yes, that would be great if a new version could be released - I may manually patch the code on my server in the meantime.

**Comment by lukebakken** [comment]
@rbro we hope to release an RC of `2.8.0` this week.

**Comment by rbro** [comment]
Thank you for your help and for pushing the release.  I confirm this issue was resolved in 2.8.0.  I am having a separate issue with check_heartbeat starting in 2.8.0, but I'll submit a separate ticket about that issue.
